### PR TITLE
Setup CI

### PR DESCRIPTION
**Pull request by DigitalBrains1**
This also establishes a proper `cabal.project` and fixes compilation with Clash 1.8.1, although there are currently still bugs in that. Before we release `clash-cores`, we need to either fix the following bugs, release Clash 1.10.0 and use that, or disable the problematic cores:
 * https://github.com/clash-lang/clash-compiler/issues/2806
 * https://github.com/clash-lang/clash-compiler/issues/2807
 * https://github.com/clash-lang/clash-cores/issues/9

It was not enough to use a global unchanging `cabal.project` and amend that with a `cabal.project.local` in CI: it did not allow altering the things that needed to be altered. So I had to overwrite part of `cabal.project` itself.

### CODE REVIEW

**Comment by DigitalBrains1** [comment]
Finally, let's do some bikeshedding. I moved this to its own message so it doesn't detract from the main points in the cover letter.

The following stanza illustrates the current multi-coloured shed in all its glory:
```cabal
test-suite unittests
  type: exitcode-stdio-1.0
```
Here we decided to simply write multiple words together (I think this was a Dunglish issue where a native Dutch speaker incorrectly writes multiple words, unit tests, as a single word, unittests). However, Cabal can't make up its mind: it says `test-suite`, combining multiple words with a dash, and `exitcode` where it just strings multiple words together. Or, rather, two words together. It then has another dash because of course.¹

However, it _does_ explicitly spell `test-suite`, which makes `clash-testsuite` rather weird because you're mixing multiple spellings in the same `.cabal` file. Here, I opted to call the test suite package `clash-cores-test-suite` which is quite a mouth full and has lots of dashes.

And another bikeshed. I dislike the `.yml` extension for YAML files. The only reason for not using `.yaml` seems to be compatibility with filename restrictions that haven't existed for three decades in any of the platforms we use. Currently, I do use `.gitlab-ci.yml` because GitLab needs configuration to change that name, but I use `.yaml` for the GitLab include files as the doc explicitly says `.yml` and `.yaml` are the two allowed extensions. Frankly, I'm inclined to configure the GitLab repo to look for `.gitlab-ci.yaml`, but I kept it at the default for now.

What do people think of the spellings? `test-suite`? `.yaml`? Let the bikeshedding beginnnnn!


¹  We could "fix" this inconsistency by bumping `cabal-version` to 3.8, where the `type` property becomes the optional default, so we could remove that :-D.

**Comment by martijnbastiaan** [comment]
Bikeshedding:

* `yaml`
* `testsuite`: `integration-tests`?
* `unittests`: `unit-tests` though this one pains me because we already Dunglish it _everywhere_.

File: test-suite/Main.hs [hunk 1]
```diff
@@ -41,30 +41,23 @@ ghcVersion3 =
...
-    Just dir -> pure dir
-    Nothing -> case os of
-      "mingw32" -> pure "C:/cabal/store" -- default ghcup location
-      _ ->  (<> "/.cabal/store") <$> getEnv "HOME"
+  -- ch <- lookupEnv "store_dir"
+  -- storeDir <- case ch of
+  --   Just dir -> pure dir
+  --   Nothing -> case os of
+  --     "mingw32" -> pure "C:/cabal/store" -- default ghcup location
+  --     _ ->  (<> "/.cabal/store") <$> getEnv "HOME"
```
Comment (martijnbastiaan) [inline test-suite/Main.hs:57]: `dd`?

**Review by martijnbastiaan — APPROVED** [review]

**Comment by DigitalBrains1** [comment]
Quote with reply:

Bikeshedding:

* `yaml`
    yes. I'll make it `.gitlab-ci.yaml` and configure GitLab accordingly
* `testsuite`: `integration-tests`? `hdl-tests`?
    `hdl-tests` makes sense to me
* `unittests`: `unit-tests` though this one pains me because we already Dunglish it _everywhere_.
    I erred on the  side of copy-pasting, but I do like internal consistency, I like `unit-tests`  actually. And `doc-tests`, right?

**Comment by martijnbastiaan** [comment]
Yeah `doctest` seems to have become one word.

**Pull request merged**